nixos/modules/services/continuous-integration/gitea-actions-runner.nix

```diff
@@ -0,0 +1,176 @@
+            type = listOf str;
+            example = literalExpression ''
+              [
+                "host:host"
+                "nixos:docker://nixos/nix:latest"
+                "ubuntu-rolling:docker://ubuntu:rolling"
+                "ubuntu-lts:docker//ubuntu:latest"
+              ]
+            '';
+            description = lib.mdDoc ''
+              Labels used to map jobs to their runtime environment.
+            '';
+            apply = concatStringsSep ",";
+          };
+        };
+      });
+    };
+  };
+
+  config = mkIf (cfg.instances != {}) {
+    assertions = let
+      tokenXorTokenFile = instance:
+        (instance.token == null && instance.tokenFile != null) ||
+        (instance.token != null && instance.tokenFile == null);
+    in  [ {
+      assertion = any tokenXorTokenFile (attrValues cfg.instances);
+      message = "Instances of gitea-actions-runner can have `token` or `tokenFile`, not both.";
+    } ];
+
+    systemd.services = let
+      mkRunnerService = name: instance:
+        nameValuePair "gitea-runner-${escapeSystemdPath name}" {
+          inherit (instance) enable;
+          description = "Gitea Actions Runner";
+          after = [
+            "network-online.target"
+          ] ++ optionals (usesDocker) [
+            "docker.service"
+          ] ++ optionals (usesPodman) [
+            "podman.service"
+          ];
+          wantedBy = [
+            "multi-user.target"
+          ];
+          environment = optionalAttrs (instance.token != null) {
+            TOKEN = "${instance.token}";
+          } // optionalAttrs (usesPodman) {
+            DOCKER_HOST = "unix:///run/podman/podman.sock";
+          };
+          path = with pkgs; [
+            coreutils
+          ];
+          serviceConfig = {
+            DynamicUser = true;
+            User = "gitea-runner";
+            StateDirectory = "gitea-runner";
+            WorkingDirectory = "-/var/lib/gitea-runner/${name}";
+            ExecStartPre = pkgs.writeShellScript "gitea-register-runner-${name}" ''
+              export INSTANCE_DIR="$STATE_DIRECTORY/${name}"
+              test -f "$INSTANCE_DIR/.runner" && exit 0
+              mkdir -vp "$INSTANCE_DIR"
+              cd "$INSTANCE_DIR"
+              ${cfg.package}/bin/act_runner register --no-interactive \
+                --instance ${escapeShellArg instance.url} \
+                --token "$TOKEN" \
+                --name ${escapeShellArg instance.name} \
+                --labels ${escapeShellArg instance.labels}
+            '';
+            ExecStart = "${cfg.package}/bin/act_runner daemon";
+            SupplementaryGroups = optionals (usesDocker) [
+              "docker"
+            ] ++ optionals (usesPodman) [
+              "podman"
+            ];
+          } // optionalAttrs (instance.tokenFile != null) {
+            LoadCredential = "tokenenv:${instance.tokenFile}";
+            EnvironmentFile = "$CREDENTIALS_DIRECTORY/tokenenv";
```

Comment:
>  $TOKEN isn't available in ExecStartPre when using this.

I kinda expected that to be fixed by now. https://github.com/systemd/systemd/issues/19604

Should be part of systemd v252.

I have no strong opinion about using `LoadCredential`, since what it transports is a single use token.

Reply:
Tested in a nixos-shell and could not get it to work.
No idea why.

```
[root@nixos:~]# systemctl --version
systemd 253 (253.2)
+PAM +AUDIT -SELINUX +APPARMOR +IMA +SMACK +SECCOMP +GCRYPT -GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY +P11KIT -QRENCODE +TPM2 +BZIP2 +LZ4 +XZ +ZLIB +ZSTD +BPF_FRAMEWORK -XKBCOMMON +UTMP -SYSVINIT default-hierarchy=unified
```